### PR TITLE
Writeable IPNS

### PR DESCRIPTION
**Pull request by whyrusleeping**
Dont CR this if you dont have lots of free time. Its just up here for regression tests

### CODE REVIEW

File: core/core.go [hunk 1]
```diff
@@ -288,6 +301,13 @@ func (n *IpfsNode) teardown() error {
 		}
 	}
 
+	if n.Blocks != nil {
+		addCloser(n.Blocks)
+	}
+	if n.IpnsFs != nil {
+		addCloser(n.IpnsFs)
+	}
```
Comment (jbenet) [inline core/core.go]: you can just call 

```go
addCloser(n.Blocks)
addCloser(n.IpnsFs)
```
all it does is check for nil

File: fuse/ipns/ipns_test.go [hunk 2]
```diff
@@ -57,6 +65,38 @@ func writeFileData(t *testing.T, data []byte, path string) []byte {
...
+	_, err := os.Stat(path)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func closeMount(mnt *fstest.Mount) {
+	if err := recover(); err != nil {
+		log.Error("Recovered panic")
+		log.Error(err)
```
Comment (jbenet) [inline fuse/ipns/ipns_test.go]: should this `t.Fatal()`?

File: fuse/ipns/ipns_test.go [hunk 3]
```diff
@@ -80,17 +127,29 @@ func setupIpnsTest(t *testing.T, node *core.IpfsNode) (*core.IpfsNode, *fstest.M
 	return node, mnt
 }
 
+func TestIpnsLocalLink(t *testing.T) {
```
Comment (jbenet) [inline fuse/ipns/ipns_test.go]: `TestIpnsLocalLinkExists` (this test doesnt really test it works)

File: fuse/ipns/ipns_test.go [hunk 4]
```diff
@@ -216,8 +306,42 @@ func TestAppendFile(t *testing.T) {
 	}
 }
 
+func TestConcurrentWrites(t *testing.T) {
```
Comment (jbenet) [inline fuse/ipns/ipns_test.go]: maybe `TestConcurrentWritesToDifferentFiles` ? i expected it to write to same file

File: fuse/ipns/ipns_test.go [hunk 5]
```diff
@@ -216,8 +306,42 @@ func TestAppendFile(t *testing.T) {
...
+	wg := sync.WaitGroup{}
+	for i := 0; i < nactors; i++ {
+		data[i] = make([][]byte, filesPerActor)
+		wg.Add(1)
+		go func(n int) {
+			defer wg.Done()
+			for j := 0; j < filesPerActor; j++ {
+				out := writeFile(t, fileSize, mnt.Dir+fmt.Sprintf("/local/%dFILE%d", n, j))
+				data[n][j] = out
+			}
```
Comment (jbenet) [inline fuse/ipns/ipns_test.go]: maybe loop a few times times? write to the same files say 10 times

File: fuse/ipns/ipns_unix.go [hunk 6]
```diff
@@ -8,28 +8,22 @@ import (
...
-	path "github.com/jbenet/go-ipfs/path"
 	ft "github.com/jbenet/go-ipfs/unixfs"
-	uio "github.com/jbenet/go-ipfs/unixfs/io"
-	ftpb "github.com/jbenet/go-ipfs/unixfs/pb"
 	u "github.com/jbenet/go-ipfs/util"
-	lgbl "github.com/jbenet/go-ipfs/util/eventlog/loggables"
 )
 
-const IpnsReadonly = true
+var DAGSERV dag.DAGService
```
Comment (jbenet) [inline fuse/ipns/ipns_unix.go]: `DagService` ?

Comment (jbenet) [inline fuse/ipns/ipns_unix.go]: also, why is this here? why is it a static var??

File: fuse/ipns/ipns_unix.go [hunk 7]
```diff
@@ -168,8 +154,27 @@ func (s *Root) Lookup(ctx context.Context, name string) (fs.Node, error) {
...
+	for _, kr := range r.Roots {
+		err := kr.Publish(r.Ipfs.Context())
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Forget is called when the filesystem is unmounted. probably.
```
Comment (jbenet) [inline fuse/ipns/ipns_unix.go]: expand on the probably. when may it happen?

File: fuse/ipns/ipns_unix.go [hunk 8]
```diff
@@ -211,91 +217,85 @@ type Node struct {
...
+// Attr returns the attributes of a given node.
+func (fi *File) Attr() fuse.Attr {
+	log.Debug("File Attr")
+	size, err := fi.fi.Size()
+	if err != nil {
+		log.Critical("Failed to get file size: %s", err)
+	}
+	return fuse.Attr{
+		Mode: os.FileMode(0666),
+		Size: uint64(size),
```
Comment (jbenet) [inline fuse/ipns/ipns_unix.go]: `Blocks` arg ok to go?

File: fuse/ipns/ipns_unix.go [hunk 9]
```diff
@@ -211,91 +217,85 @@ type Node struct {
...
-		child.nsRoot = n
-	} else {
-		child.nsRoot = n.nsRoot
+	switch child := child.(type) {
+	case *nsfs.Directory:
+		return &Directory{child}, nil
+	case nsfs.File:
+		return &File{fi: child}, nil
+	default:
+		panic("system has proven to be insane")
```
Comment (jbenet) [inline fuse/ipns/ipns_unix.go]: meaningful error for user. "unexpected error: ipns dir lookup type failed"

File: fuse/ipns/ipns_unix.go [hunk 10]
```diff
@@ -304,300 +304,145 @@ func (s *Node) ReadDirAll(ctx context.Context) ([]fuse.Dirent, error) {
 	return nil, fuse.ENOENT
 }
 
-func (s *Node) Read(ctx context.Context, req *fuse.ReadRequest, resp *fuse.ReadResponse) error {
-	k, err := s.Nd.Key()
+func (fi *File) Read(ctx context.Context, req *fuse.ReadRequest, resp *fuse.ReadResponse) error {
```
Comment (jbenet) [inline fuse/ipns/ipns_unix.go]: func ignores `ctx`-- should this be return early if cancelled?

File: fuse/ipns/ipns_unix.go [hunk 11]
```diff
@@ -304,300 +304,145 @@ func (s *Node) ReadDirAll(ctx context.Context) ([]fuse.Dirent, error) {
...
-	if n.dagMod == nil {
-		// Create a DagModifier to allow us to change the existing dag node
-		dmod, err := uio.NewDagModifier(n.Nd, n.Ipfs.DAG, chunk.DefaultSplitter)
-		if err != nil {
-			return err
-		}
-		n.dagMod = dmod
-	}
-	wrote, err := n.dagMod.WriteAt(req.Data, uint64(req.Offset))
+func (fi *File) Write(ctx context.Context, req *fuse.WriteRequest, resp *fuse.WriteResponse) error {
```
Comment (jbenet) [inline fuse/ipns/ipns_unix.go]: func ignores `ctx`?

File: fuse/ipns/ipns_unix.go [hunk 12]
```diff
@@ -304,300 +304,145 @@ func (s *Node) ReadDirAll(ctx context.Context) ([]fuse.Dirent, error) {
...
-		return
-	}
-	root := n.nsRoot
-	if root == nil {
-		root = n
-	}
-
-	root.repub.Publish <- struct{}{}
+func (fi *File) Flush(ctx context.Context, req *fuse.FlushRequest) error {
+	return fi.fi.Close()
```
Comment (jbenet) [inline fuse/ipns/ipns_unix.go]: `Close` certainly flushes?

File: fuse/ipns/ipns_unix.go [hunk 13]
```diff
@@ -304,300 +304,145 @@ func (s *Node) ReadDirAll(ctx context.Context) ([]fuse.Dirent, error) {
...
 	}
+}
 
-	if n.parent != nil {
-		err := n.parent.update(n.name, nnode)
-		if err != nil {
-			log.Criticalf("Error updating node: %s", err)
-			return nil, err
-		}
+func (dir *Directory) Mkdir(ctx context.Context, req *fuse.MkdirRequest) (fs.Node, error) {
```
Comment (jbenet) [inline fuse/ipns/ipns_unix.go]: perhaps the `Directory` and `File` objects should be unexported, no?

File: fuse/ipns/ipns_unix.go [hunk 14]
```diff
@@ -304,300 +304,145 @@ func (s *Node) ReadDirAll(ctx context.Context) ([]fuse.Dirent, error) {
...
-	return child, nil
+	return &Directory{child}, nil
 }
 
-func (n *Node) Open(ctx context.Context, req *fuse.OpenRequest, resp *fuse.OpenResponse) (fs.Handle, error) {
+func (fi *File) Open(ctx context.Context, req *fuse.OpenRequest, resp *fuse.OpenResponse) (fs.Handle, error) {
 	//log.Debug("[%s] Received open request! flags = %s", n.name, req.Flags.String())
 	//TODO: check open flags and truncate if necessary
 	if req.Flags&fuse.OpenTruncate != 0 {
 		log.Warning("Need to truncate file!")
```
Comment (jbenet) [inline fuse/ipns/ipns_unix.go]: drop from warning to Debug?

File: fuse/ipns/ipns_unix.go [hunk 15]
```diff
@@ -304,300 +304,145 @@ func (s *Node) ReadDirAll(ctx context.Context) ([]fuse.Dirent, error) {
...
-		n.cached = nil
-		n.Nd = &mdag.Node{Data: ft.FilePBData(nil, 0)}
+		err := fi.fi.Truncate(0)
+		if err != nil {
+			return nil, err
+		}
 	} else if req.Flags&fuse.OpenAppend != 0 {
 		log.Warning("Need to append to file!")
```
Comment (jbenet) [inline fuse/ipns/ipns_unix.go]: drop from Warnign to Debug?

File: fuse/readonly/readonly_unix.go [hunk 16]
```diff
@@ -170,7 +170,7 @@ func (s *Node) Read(ctx context.Context, req *fuse.ReadRequest, resp *fuse.ReadR
 		return err
 	}
 
-	buf := resp.Data[:min(req.Size, int(r.Size()))]
+	buf := resp.Data[:min(req.Size, int(r.Size()-req.Offset))]
```
Comment (jbenet) [inline fuse/readonly/readonly_unix.go]: maybe this should be its own PR

**Comment by jbenet** [comment]
This is a very large PR-- perhaps worth splitting out all the little things that are preparation into other PR? Easier to look over twice.

File: importer/helpers/dagbuilder.go [hunk 17]
```diff
@@ -78,6 +78,11 @@ func (db *DagBuilderHelper) Next() []byte {
 	return d
 }
 
+// GetDagServ returns the dagservice object this Helper is using
+func (db *DagBuilderHelper) GetDagServ() dag.DAGService {
+	return db.dserv
+}
```
Comment (jbenet) [inline importer/helpers/dagbuilder.go]: Prep PR?

File: importer/helpers/dagbuilder.go [hunk 18]
```diff
@@ -86,7 +91,7 @@ func (db *DagBuilderHelper) FillNodeLayer(node *UnixfsNode) error {
 
 	// while we have room AND we're not done
 	for node.NumChildren() < db.maxlinks && !db.Done() {
-		child := NewUnixfsNode()
+		child := NewUnixfsBlock()
```
Comment (jbenet) [inline importer/helpers/dagbuilder.go]: Prep PR?

File: importer/helpers/helpers.go [hunk 19]
```diff
@@ -38,16 +38,35 @@ var ErrSizeLimitExceeded = fmt.Errorf("object size limit exceeded")
...
+	if err != nil {
+		return nil, err
+	}
+
+	return &UnixfsNode{
+		node: nd,
+		ufmt: mb,
+	}, nil
+}
+
```
Comment (jbenet) [inline importer/helpers/helpers.go]: Prep PR?

File: importer/helpers/helpers.go [hunk 20]
```diff
@@ -38,16 +38,35 @@ var ErrSizeLimitExceeded = fmt.Errorf("object size limit exceeded")
...
 
 func NewUnixfsNode() *UnixfsNode {
 	return &UnixfsNode{
 		node: new(dag.Node),
-		ufmt: new(ft.MultiBlock),
+		ufmt: &ft.FSNode{Type: ft.TFile},
 	}
 }
 
+func NewUnixfsBlock() *UnixfsNode {
```
Comment (jbenet) [inline importer/helpers/helpers.go]: Maybe document why `NewUnixfsBlock` differs from `NewUnixfsNode` and `NewUnixfsNodeFromDag`?

File: importer/helpers/helpers.go [hunk 21]
```diff
@@ -38,16 +38,35 @@ var ErrSizeLimitExceeded = fmt.Errorf("object size limit exceeded")
...
+	return &UnixfsNode{
+		node: new(dag.Node),
+		ufmt: &ft.FSNode{Type: ft.TRaw},
```
Comment (jbenet) [inline importer/helpers/helpers.go]: If i'm reading this correctly... why is there a need for differentiating `ft.TRaw` and `ft.TFile`? 

Making them different types breaks the amazing composability that you get from them being the same. No longer can one file be embedded in another _completely_.

File: importer/helpers/helpers.go [hunk 22]
```diff
@@ -83,7 +102,13 @@ func (n *UnixfsNode) AddChild(child *UnixfsNode, db *DagBuilderHelper) error {
...
 }
 
-func (n *UnixfsNode) setData(data []byte) {
+// Removes the child node at the given index
+func (n *UnixfsNode) RemoveChild(index int) {
+	n.ufmt.RemoveBlockSize(index)
+	n.node.Links = append(n.node.Links[:index], n.node.Links[index+1:]...)
+}
+
+func (n *UnixfsNode) SetData(data []byte) {
```
Comment (jbenet) [inline importer/helpers/helpers.go]: Prep PR

File: importer/trickle/trickle_test.go [hunk 23]
```diff
@@ -441,3 +442,166 @@ func TestSeekingConsistency(t *testing.T) {
...
 	}
 }
+
+func TestAppend(t *testing.T) {
+	nbytes := int64(128 * 1024)
+	should := make([]byte, nbytes)
+	u.NewTimeSeededRand().Read(should)
+
+	// Reader for half the bytes
+	read := bytes.NewReader(should[:64*1024])
```
Comment (jbenet) [inline importer/trickle/trickle_test.go]: maybe make `64:1024` be `nbytes/2` -- more resistant to change

File: importer/trickle/trickle_test.go [hunk 24]
```diff
@@ -441,3 +442,166 @@ func TestSeekingConsistency(t *testing.T) {
...
+
+func TestMultipleAppends(t *testing.T) {
+	ds := mdtest.Mock(t)
+
+	// TODO: fix small size appends and make this number bigger
+	nbytes := int64(1000)
+	should := make([]byte, nbytes)
+	u.NewTimeSeededRand().Read(should)
+
+	// Reader for half the bytes
```
Comment (jbenet) [inline importer/trickle/trickle_test.go]: is this comment supposed to be here?

File: importer/trickle/trickle_test.go [hunk 25]
```diff
@@ -441,3 +442,166 @@ func TestSeekingConsistency(t *testing.T) {
...
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestMultipleAppends(t *testing.T) {
+	ds := mdtest.Mock(t)
+
+	// TODO: fix small size appends and make this number bigger
+	nbytes := int64(1000)
```
Comment (jbenet) [inline importer/trickle/trickle_test.go]: should this number be bigger now?

File: importer/trickle/trickledag.go [hunk 26]
```diff
@@ -56,3 +56,181 @@ func fillTrickleRec(db *h.DagBuilderHelper, node *h.UnixfsNode, depth int) error
...
+
+	// Get depth of this 'tree'
+	n, j := trickleDepthInfo(base, db.Maxlinks())
+	if n == 0 {
+		// If direct blocks not filled...
+		err := db.FillNodeLayer(ufsn)
+		if err != nil {
+			return nil, err
+		}
+		n++
```
Comment (jbenet) [inline importer/trickle/trickledag.go]: feels like a stray `n++` here, doesnt hurt to put it at the top of the if statement for readability.

File: importer/trickle/trickledag.go [hunk 27]
```diff
@@ -56,3 +56,181 @@ func fillTrickleRec(db *h.DagBuilderHelper, node *h.UnixfsNode, depth int) error
...
+			}
+
+			err = ufsn.AddChild(next, db)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return ufsn.GetDagNode()
```
Comment (jbenet) [inline importer/trickle/trickledag.go]: maybe i'm just tired but i dont understand what the algorithm is doing. i roughly know what it's supposed to end up like, but maybe more doc would help very much.

**Comment by jbenet** [comment]
I think the TrickleDAG algo is very error prone-- lots of special casing and edge casing-- can the code be simpler? more expressive? more functions may help.

File: ipnsfs/dir.go [hunk 28]
```diff
@@ -0,0 +1,340 @@
+package ipnsfs
```
Comment (jbenet) [inline ipnsfs/dir.go]: so `ipnsfs` is a pkg sort of like `unixfs`?

File: ipnsfs/dir.go [hunk 29]
```diff
@@ -0,0 +1,340 @@
...
+	}
+	return dir.Open(tpath[1:], mode)
+}
+
+// consider combining into a single method...
+type childCloser interface {
+	closeChild(string, *dag.Node) error
+}
+
+func (d *Directory) closeChild(name string, nd *dag.Node) error {
```
Comment (jbenet) [inline ipnsfs/dir.go]: maybe document what this close procedure is doing and why? not clear readily

File: ipnsfs/dir.go [hunk 30]
```diff
@@ -0,0 +1,340 @@
...
+		d.lock.Unlock()
+		return err
+	}
+
+	err = d.node.AddNodeLinkClean(name, nd)
+	if err != nil {
+		d.lock.Unlock()
+		return err
+	}
+	d.lock.Unlock()
```
Comment (jbenet) [inline ipnsfs/dir.go]: a pattern i like for these undeferrable lock/unlocks is:

```go
lockedFunc := func() error {
  // do stuff that needs to happen while locked.
}

d.lock.Lock()
err := lockedFunc()
d.lock.Unlock()
if err !=  nil {
  return err
}
```

File: ipnsfs/dir.go [hunk 31]
```diff
@@ -0,0 +1,340 @@
...
+				nfi, err := NewFile(name, nd, d, d.fs)
+				if err != nil {
+					return nil, err
+				}
+				d.files[name] = nfi
+				return nfi, nil
+			case ufspb.Data_Metadata:
+				panic("NOT YET IMPLEMENTED")
+			default:
+				panic("NO!")
```
Comment (jbenet) [inline ipnsfs/dir.go]: more meaningful errors

File: ipnsfs/dir.go [hunk 32]
```diff
@@ -0,0 +1,340 @@
...
+			case ufspb.Data_Directory:
+				ndir := NewDirectory(name, nd, d, d.fs)
+				d.childDirs[name] = ndir
+				return ndir, nil
+			case ufspb.Data_File:
+				return nil, fmt.Errorf("%s is not a directory", name)
+			case ufspb.Data_Metadata:
+				panic("NOT YET IMPLEMENTED")
+			default:
+				panic("NO!")
```
Comment (jbenet) [inline ipnsfs/dir.go]: more meaningful errors

File: ipnsfs/dir.go (hunk 28 above)
Comment (jbenet) [inline ipnsfs/dir.go]: i cant check for correctness until i understand more all the things this package is supposed to handle. maybe write a package doc?

File: ipnsfs/system.go [hunk 33]
```diff
@@ -0,0 +1,285 @@
+package ipnsfs
```
Comment (jbenet) [inline ipnsfs/system.go]: it seems to me like most of this package can be a generalized "MutableFS" or something, and only the top is ipns specific-- i may be missing something

File: ipnsfs/system.go [hunk 34]
```diff
@@ -0,0 +1,285 @@
...
+		if err != nil {
+			return nil, err
+		}
+
+		root, err := fs.NewKeyRoot(ctx, k)
+		if err != nil {
+			return nil, err
+		}
+		roots[u.Key(pkh).Pretty()] = root
+	}
```
Comment (jbenet) [inline ipnsfs/system.go]: what if we generate a key mid-run-- can this handle that? (generating all the roots beforehand makes me ask)

File: ipnsfs/system.go [hunk 35]
```diff
@@ -0,0 +1,285 @@
...
+
+	root.node = mnode
+
+	root.repub = NewRepublisher(root, time.Millisecond*300, time.Second*3)
+	go root.repub.Run(parent)
+
+	pbn, err := ft.FromBytes(mnode.Data)
+	if err != nil {
+		log.Error("IPNS pointer was not unixfs node")
```
Comment (jbenet) [inline ipnsfs/system.go]: why does it have to be? -- ipns may point to any datastructure

File: ipnsfs/system.go [hunk 36]
```diff
@@ -0,0 +1,285 @@
...
+	case ft.TDirectory:
+		root.val = NewDirectory(pointsTo.B58String(), mnode, root, fs)
+	case ft.TFile, ft.TMetadata, ft.TRaw:
+		fi, err := NewFile(pointsTo.B58String(), mnode, root, fs)
+		if err != nil {
+			return nil, err
+		}
+		root.val = fi
+	default:
+		panic("unrecognized! (NYI)")
```
Comment (jbenet) [inline ipnsfs/system.go]: more meaningful error message

File: ipnsfs/system.go [hunk 37]
```diff
@@ -0,0 +1,285 @@
...
+	k, err := kr.fs.dserv.Add(nd)
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("Publishing!")
+	return kr.fs.nsys.Publish(ctx, kr.key, k)
+}
+
+// Republisher manages when to publish the ipns entry associated with a given key
```
Comment (jbenet) [inline ipnsfs/system.go]: maybe expand on why this is necessary. if i didn't remember i'd be confused.

File: ipnsfs/system.go [hunk 38]
```diff
@@ -0,0 +1,285 @@
...
+	for {
+		select {
+		case <-np.Publish:
+			quick := time.After(np.TimeoutShort)
+			longer := time.After(np.TimeoutLong)
+
+		wait:
+			select {
+			case <-quick:
+			case <-longer:
```
Comment (jbenet) [inline ipnsfs/system.go]: maybe this order?

```go
select {
case <-ctx.Done():
  ...
case <-np.Publish:
  ...
case <-quick:
case <-longer:
}
```
so it's clear the last two just fall through out of the select.  when first reading this i somehow took it to mean all three (quick, longer, ctx.Done) returned-- which is not correct Go, but the code sort of leads me there

File: ipnsfs/system_test.go [hunk 39]
```diff
@@ -0,0 +1,92 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	t.Log(nb)
+
+	if !bytes.Equal(nb, data) {
+		t.Fatal("data not the same after closing down fs")
+	}
+}
```
Comment (jbenet) [inline ipnsfs/system_test.go]: this package is _very_ complex. it needs pretty robust tests on its own. this basic test is not enough.

it would be useful to have large randomized tests in _all_ packages of this sort. things like:

```
for 10 times {
  for 1000 dags {
    write dag out
    read dag in
    compare
  }
}
```

Comment (jbenet) [inline ipnsfs/system_test.go]: i think the bitswap tests are a good example of full testing of all the different parameters. and the `epic` test too.

File: unixfs/mod/dagmodifier.go [hunk 40]
```diff
@@ -0,0 +1,428 @@
+package mod
```
Comment (jbenet) [inline unixfs/mod/dagmodifier.go]: did the dag modifier change at all? i cant tell. it would be useful to put this in its own pr and in two separate commits.

**Comment by jbenet** [comment]
Major note: there are many different PRs in here:

- unixfs offset fix 
- the mutable FS thing (most of the ipnsfs pkg)
- ipns mount implementation
- importer raw node type? (why is this needed)
- trickle dag
- dag modifier move

It would be really useful to examine, test, + merge each in isolation, let's split them.

**Comment by cryptix** [comment]
:+1: on splitting up into multiple PRs

File: unixfs/mod/dagmodifier.go (hunk 40 above)
Comment (whyrusleeping) [inline unixfs/mod/dagmodifier.go]: Yeah, git diff apparently doesnt recognize them as the same code anymore, or else it would have shown up as a rename.

File: ipnsfs/system.go (hunk 38 above)
Comment (whyrusleeping) [inline ipnsfs/system.go]: Yeah, that makes sense. (you wrote this btw :P )

File: ipnsfs/system.go (hunk 35 above)
Comment (whyrusleeping) [inline ipnsfs/system.go]: this has to be for now since we dont have a defined way of reading other dag formats.

File: ipnsfs/system.go (hunk 33 above)
Comment (whyrusleeping) [inline ipnsfs/system.go]: agreed. `ipnsfs` was just a quick working name i gave it so i could get to coding. I also dont know if I like it being in the root namespace, but not sure where else it should go.

File: fuse/ipns/ipns_unix.go (hunk 8 above)
Comment (whyrusleeping) [inline fuse/ipns/ipns_unix.go]: Yeah, it doesnt really mean anything. All it does (as far as i know) is fill out a field in the stat output. And we dont have an easy way of counting blocks

File: fuse/ipns/ipns_unix.go (hunk 7 above)
Comment (whyrusleeping) [inline fuse/ipns/ipns_unix.go]: Yeah, i might actually remove this method, the fuse documentation is very unclear about when this is called and what it signifies.

File: fuse/ipns/ipns_unix.go (hunk 6 above)
Comment (whyrusleeping) [inline fuse/ipns/ipns_unix.go]: whoops. debugging, thus the all caps (I name things ugly when they are supposed to temporary)

File: fuse/ipns/ipns_test.go (hunk 5 above)
Comment (whyrusleeping) [inline fuse/ipns/ipns_test.go]: good idea!

File: core/core.go (hunk 1 above)
Comment (whyrusleeping) [inline core/core.go]: nope: https://github.com/jbenet/go-ipfs/issues/855

File: fuse/readonly/readonly_unix.go (hunk 16 above)
Comment (whyrusleeping) [inline fuse/readonly/readonly_unix.go]: #864

File: importer/helpers/dagbuilder.go (hunk 17 above)
Comment (whyrusleeping) [inline importer/helpers/dagbuilder.go]: #865

File: importer/helpers/helpers.go (hunk 21 above)
Comment (whyrusleeping) [inline importer/helpers/helpers.go]: well... we havent ever been able to do that. We've gotten lucky that we havent run into a situation where its bitten us.
